lib/datadog/statsd/sender.rb

```diff
@@ -1,159 +1,130 @@
 # frozen_string_literal: true
 
 module Datadog
+  class FlushQueue < Queue
+  end
+  class CloseQueue < Queue
+  end
   class Statsd
-    # Sender is using a companion thread to flush and pack messages
+    # Sender is using a background thread to flush and pack messages
     # in a `MessageBuffer`.
     # The communication with this thread is done using a `Queue`.
     # If the thread is dead, it is starting a new one to avoid having a blocked
-    # Sender with no companion thread to communicate with (most of the time, having
-    # a dead companion thread means that a fork just happened and that we are
-    # running in the child process).
+    # Sender with no background thread to communicate with (most of the time,
+    # having a dead background thread means that a fork just happened and that we
+    # are running in the child process).
     class Sender
       CLOSEABLE_QUEUES = Queue.instance_methods.include?(:close)
 
       def initialize(message_buffer, logger: nil)
         @message_buffer = message_buffer
         @logger = logger
-        @mx = Mutex.new
-      end
 
-      def flush(sync: false)
-        # keep a copy around in case another thread is calling #stop while this method is running
-        current_message_queue = message_queue
-
-        # don't try to flush if there is no message_queue instantiated or
-        # no companion thread running
-        if !current_message_queue
-          @logger.debug { "Statsd: can't flush: no message queue ready" } if @logger
-          return
-        end
-        if !sender_thread.alive?
-          @logger.debug { "Statsd: can't flush: no sender_thread alive" } if @logger
-          return
-        end
+        # communication and synchronization with the background thread
+        # @mux is also used to not having multiple threads fighting for
+        # closing the Sender or creating a new background thread
+        @channel = Queue.new
+        @mux = Mutex.new
+
+        @is_closed = false
 
-        current_message_queue.push(:flush)
-        rendez_vous if sync
+        # start background thread immediately
+        @sender_thread = Thread.new(&method(:send_loop))
       end
 
-      def rendez_vous
-        # could happen if #start hasn't be called
-        return unless message_queue
-
-        # Initialize and get the thread's sync queue
-        queue = (Thread.current[:statsd_sync_queue] ||= Queue.new)
```

Comment:
We lost this caching-the-queue behavior in the refactoring, which doesn't seem like an issue, but just doublechecking by asking if this is ok from a performance pov (I actually am not sure how expensive it is to create queues, probably not a lot)